src/ber/util.rs

```diff
@@ -0,0 +1,140 @@
+use super::{DeserializeError, Length, Serialize, Tag};
+use tinyvec::TinyVec;
+
+pub trait VecExt {
+    fn write(&mut self, slice: &[u8]);
+    fn write_byte(&mut self, byte: u8);
+}
+
+impl VecExt for Vec<u8> {
+    fn write(&mut self, slice: &[u8]) {
+        self.extend_from_slice(slice);
+    }
+    fn write_byte(&mut self, byte: u8) {
+        self.push(byte);
+    }
+}
+
+impl VecExt for TinyVec<[u8; 32]> {
+    fn write(&mut self, slice: &[u8]) {
+        self.extend_from_slice(slice);
+    }
+    fn write_byte(&mut self, byte: u8) {
+        self.push(byte);
+    }
+}
+
+pub trait ReadExt {
```

Comment:
given that the error is same for all, we could use an associated type instead. not necessary but just a thought

Reply:
Not a bad idea, but pretty low priority, might do it in a future commit once the error handling stuff is actually set up